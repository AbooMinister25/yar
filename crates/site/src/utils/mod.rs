use std::path::{Path, PathBuf};

use color_eyre::{Result, eyre::ContextCompat};
use url::Url;

pub mod fs;

/// Build permalink for a site item.
pub fn build_permalink<P: AsRef<Path>, T: AsRef<Path>>(
    path: P,
    out_dir: T,
    url: &Url,
) -> Result<Url> {
    let out = out_dir
        .as_ref()
        .file_name()
        .context("Output directory shouldn't terminate in ..")?;

    let mut components = path
        .as_ref()
        .components()
        .skip_while(|c| c.as_os_str() != out);
    components.next();
    let mut url_ending = components.collect::<PathBuf>();
    if url_ending.ends_with("index.html") {
        url_ending = url_ending
            .parent()
            .context("path doesn't have parent?")?
            .to_path_buf();
    }
    let permalink = url.join(
        url_ending
            .to_str()
            .context("Path should be valid unicode.")?,
    )?;

    Ok(permalink)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_permalink() -> Result<()> {
        let path = Path::new("site/index.html");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        let path = Path::new("site/posts/hello-world/index.html");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        let path = Path::new("site/assets/style.css");
        let out_dir = Path::new("site/");
        let url = Url::parse("https://example.com/")?;
        insta::assert_yaml_snapshot!(build_permalink(path, out_dir, &url)?);

        Ok(())
    }
}
