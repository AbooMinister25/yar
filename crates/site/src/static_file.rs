use std::{
    fs,
    path::{Path, PathBuf},
};

use color_eyre::{Result, eyre::ContextCompat};

use crate::utils::{build_permalink, fs::ensure_directory};

/// Represents a static asset. These are copied over to the resulting
/// site as-is.
#[derive(Debug)]
pub struct StaticFile {
    pub path: PathBuf,
    pub source_hash: String,
    pub out_path: PathBuf,
    pub permalink: String,
    pub content: String,
}

impl StaticFile {
    pub fn new<P: AsRef<Path>, T: AsRef<Path>, Z: AsRef<Path>>(
        path: P,
        source_hash: String,
        out_dir: T,
        root: Z,
        url: &str,
    ) -> Result<Self> {
        let out_path = out_path(&path, &out_dir, root);
        let permalink = build_permalink(&out_path, url)?;
        let content = fs::read_to_string(&path)?;

        Ok(Self {
            path: path.as_ref().to_owned(),
            source_hash,
            out_path,
            permalink,
            content,
        })
    }

    pub fn render(&self) -> Result<()> {
        ensure_directory(
            self.out_path
                .parent()
                .context("Path should have a parent")?,
        )?;
        fs::copy(&self.path, &self.out_path)?;
        Ok(())
    }
}

fn out_path<P: AsRef<Path>, T: AsRef<Path>, Z: AsRef<Path>>(
    path: P,
    out_dir: T,
    root: Z,
) -> PathBuf {
    let out_dir = out_dir.as_ref();
    let mut components = path
        .as_ref()
        .components()
        .filter(|c| !c.as_os_str().to_str().is_some_and(|s| s.starts_with("_")));

    if root.as_ref() != Path::new(".") {
        components.next();
    }

    out_dir.components().chain(components).collect::<PathBuf>()
}
