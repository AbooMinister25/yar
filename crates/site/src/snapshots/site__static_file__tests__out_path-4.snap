---
source: crates/site/src/static_file.rs
expression: path
---
public/image.png
